pkg/kubelet/logs/container_log_manager.go

```diff
@@ -213,80 +223,121 @@ func (c *containerLogManager) Clean(ctx context.Context, containerID string) err
-			// We try to recover it by reopening container log.
-			if err := c.runtimeService.ReopenContainerLog(ctx, id); err != nil {
-				klog.ErrorS(err, "Container log doesn't exist, reopen container log failed", "containerID", id, "path", path)
-				continue
-			}
-			// The container log should be recovered.
-			info, err = c.osInterface.Stat(path)
-			if err != nil {
-				klog.ErrorS(err, "Failed to stat container log after reopen", "path", path)
-				continue
-			}
+		c.queue.Add(container.GetId())
+	}
+	wait.PollInfinite(1*time.Second, func() (done bool, err error) {
+		return c.queue.Len() == 0, nil
+	})
+	klog.V(4).InfoS("Finished container log rotation sequence")
+	return nil
+}
+
+func (c *containerLogManager) processContainer(ctx context.Context, worker int) (ok bool) {
+	// Only rotate logs for running containers. Non-running containers won't
+	// generate new output, it doesn't make sense to keep an empty latest log.
+	// Note that we should not block log rotate for an error of a single container.
+	// In rotateLatestLog, there are several cases that we may
+	// lose original container log after ReopenContainerLog fails.
+	// We try to recover it by reopening container log.
+	// The container log should be recovered.
+	// Perform log rotation.
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer func() {
+		c.queue.Done(key)
+		c.queue.Forget(key)
+	}()
+	// Always default the return to true to keep the processing of Queue ongoing
+	ok = true
+	id := key.(string)
+
+	resp, err := c.runtimeService.ContainerStatus(ctx, id, false)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get container status", "worker", worker, "containerID", id)
+		return
+	}
+	if resp.GetStatus() == nil {
+		klog.ErrorS(err, "Container status is nil", "worker", worker, "containerID", id)
+		return
+	}
+	path := resp.GetStatus().GetLogPath()
+	info, err := c.osInterface.Stat(path)
+
+	if err != nil {
+		if !os.IsNotExist(err) {
+			klog.ErrorS(err, "Failed to stat container log", "worker", worker, "containerID", id, "path", path)
+			return
 		}
-		if info.Size() < c.policy.MaxSize {
-			continue
+
+		if err := c.runtimeService.ReopenContainerLog(ctx, id); err != nil {
+			klog.ErrorS(err, "Container log doesn't exist, reopen container log failed", "worker", worker, "containerID", id, "path", path)
+			return
 		}
-		// Perform log rotation.
-		if err := c.rotateLog(ctx, id, path); err != nil {
-			klog.ErrorS(err, "Failed to rotate log for container", "path", path, "containerID", id)
-			continue
+
+		info, err = c.osInterface.Stat(path)
+		if err != nil {
+			klog.ErrorS(err, "Failed to stat container log after reopen", "worker", worker, "containerID", id, "path", path)
+			return
 		}
 	}
-	return nil
+	if info.Size() < c.policy.MaxSize {
+		klog.V(4).InfoS("log file doesn't need to be rotated", "worker", worker, "containerID", id, "path", path, "currentSize", info.Size(), "maxSize", c.policy.MaxSize)
```

Comment:
Doesn't seem unreasonable, but will add a bunch of noise to the logs if verbosity increased.

Reply:
I will move this to a much higher verbosity may be ? like 7 instead of 4 ?

Reply:
sounds fair